File: doubleFL.cpp
#include <iostream>
#include <string>
#include <fstream>
#include "doubleFL.h"

using namespace std;

void createDoubleFL(string beginNode, doubleFL& doubleFlist) {
    doubleFlist.head = new DNode{beginNode, nullptr, nullptr};
    doubleFlist.tail = doubleFlist.head;
}

void addAtHead(doubleFL& doubleFlist, string data) {
    if (doubleFlist.head == nullptr) {
        doubleFlist.head = new DNode{data, nullptr, nullptr};
        doubleFlist.tail = doubleFlist.head;
        return;
    }
    
    DNode* newHead = new DNode{data, doubleFlist.head, nullptr};
    doubleFlist.head->prev = newHead;
    doubleFlist.head = newHead;
}

void addAtTail(doubleFL& doubleFlist, string data) {
    if (doubleFlist.head == nullptr) {
        doubleFlist.head = new DNode{data, nullptr, nullptr};
        doubleFlist.tail = doubleFlist.head;
        return;
    }
    
    DNode* newNode = new DNode{data, nullptr, doubleFlist.tail};
    doubleFlist.tail->next = newNode;
    doubleFlist.tail = newNode;
}

void addBeforeIndex(doubleFL& doubleFlist, string data, int index) {
    if (doubleFlist.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }
    else if (index == 0) {
        addAtHead(doubleFlist, data);
        return;
    }
    else if (index < 0) {
        cout << "Неверный индекс!\n";
        return;
    }

    int i = 0;
    DNode* curElem = doubleFlist.head;

    for (; i != index && curElem != nullptr; i++) {
        curElem = curElem->next;
    }

    if (curElem == nullptr) {
        cout << "Элемента с таким индексом не существует!\n";
        return;
    }

    DNode* newNode = new DNode{data, curElem, curElem->prev};
    curElem->prev->next = newNode;
    curElem->prev = newNode;
}

void addAfterIndex(doubleFL& doubleFlist, string data, int index) {
    if (doubleFlist.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }
    else if (index < 0) {
        cout << "Неверный индекс!\n";
        return;
    }

    DNode* curElem = doubleFlist.head;
    int i = 0;

    for (; i != index && curElem != nullptr; i++) {
        curElem = curElem->next;
    }

    if (curElem == nullptr) {
        cout << "Элемента с таким индексом не существует!\n";
        return;
    }

    if (curElem == doubleFlist.tail) {
        addAtTail(doubleFlist, data);
        return;
    }

    DNode* newNode = new DNode{data, curElem->next, curElem};
    curElem->next->prev = newNode;
    curElem->next = newNode;
}

void deleteAtHead(doubleFL& doubleFlist) {
    if (doubleFlist.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }
    
    DNode* deleteNode = doubleFlist.head;

    if (doubleFlist.head == doubleFlist.tail) {
        doubleFlist.head = doubleFlist.tail = nullptr;
    }
    else {
        doubleFlist.head = doubleFlist.head->next;
        doubleFlist.head->prev = nullptr;
    }

    delete deleteNode;
}

void deleteAtTail(doubleFL& doubleFlist) {
    if (doubleFlist.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }
    
    DNode* deleteNode = doubleFlist.tail;

    if (doubleFlist.head == doubleFlist.tail) {
        doubleFlist.head = doubleFlist.tail = nullptr;
    }
    else {
        doubleFlist.tail->prev->next = nullptr;
        doubleFlist.tail = doubleFlist.tail->prev;
    }

    delete deleteNode;
}

void deleteBeforeIndex(doubleFL& doubleFlist, int index) {
    if (doubleFlist.head == nullptr || doubleFlist.head == doubleFlist.tail) {
        cout << "Удаление невозможно.\n";
        return;
    }
    else if (index == 1) {
        deleteAtHead(doubleFlist);
        return;
    }
    else if (index <= 0) {
        cout << "Неверный индекс.\n";
        return;
    }

    DNode* curElem = doubleFlist.head;
    int i = 0;

    for (; i != index && curElem != nullptr; i++) {
        curElem = curElem->next;
    }

    if (curElem == nullptr || curElem->prev == nullptr) {
        cout << "Неверный индекс.\n";
        return;
    }

    DNode* deleteNode = curElem->prev;
    
    if (deleteNode == doubleFlist.head) {
        deleteAtHead(doubleFlist);
        return;
    }

    deleteNode->prev->next = curElem;
    curElem->prev = deleteNode->prev;
    delete deleteNode;
}

void deleteAfterIndex(doubleFL& doubleFlist, int index) {
    if (doubleFlist.head == nullptr || doubleFlist.head == doubleFlist.tail) {
        cout << "Удаление невозможно.\n";
        return;
    }
    else if (index < 0) {
        cout << "Неверный индекс.\n";
        return;
    }
    
    DNode* curElem = doubleFlist.head;
    int i = 0;

    for (; i != index && curElem != nullptr; i++) {
        curElem = curElem->next;
    }

    if (curElem == nullptr || curElem == doubleFlist.tail) {
        cout << "Неверный индекс.\n";
        return;
    }

    if (curElem->next == doubleFlist.tail) {
        deleteAtTail(doubleFlist);
        return;
    }

    DNode* deleteNode = curElem->next;
    curElem->next = deleteNode->next;
    deleteNode->next->prev = curElem;
    delete deleteNode;
}

void deleteValue(doubleFL& doubleFlist, string targetValue) {
    if (doubleFlist.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }

    DNode* curElem = doubleFlist.head;
    bool isFound = false;

    while (curElem != nullptr) {
        if (curElem->value == targetValue) {
            isFound = true;
            break;
        }

        curElem = curElem->next;
    }

    if (!isFound) {
        cout << "Элемент не найден.\n";
        return;
    }

    if (curElem == doubleFlist.head) {
        deleteAtHead(doubleFlist);
        return;
    }

    if (curElem == doubleFlist.tail) {
        deleteAtTail(doubleFlist);
        return;
    }

    DNode* deleteNode = curElem;
    deleteNode->prev->next = deleteNode->next;
    deleteNode->next->prev = deleteNode->prev;
    delete deleteNode;
}

void searchValue(doubleFL& doubleFlist, string targetValue) {
    if (doubleFlist.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }

    DNode* curElem = doubleFlist.head;
    bool isFound = false;
    int index = 0;

    while (curElem != nullptr) {
        if (curElem->value == targetValue) {
            isFound = true;
            break;
        }

        curElem = curElem->next;
        index++;
    }

    if (!isFound) {
        cout << "Элемент не найден.\n";
        return;
    }
    
    cout << "Элемент " << curElem->value << " найден. Индекс: " << index << endl;
}

void printDoubleFL (DNode* doubleFlist) {
    cout << "Двусвязный список: ";

    if (doubleFlist == nullptr) {
        cout << "Список пуст\n";
        return;
    }

    while (doubleFlist != nullptr) {
        cout << doubleFlist->value << " ";
        doubleFlist = doubleFlist->next;
    }

    cout << endl;
}

File: doubleFL.go
package main

import "fmt"

type DNode struct {
    value string
    next  *DNode
    prev  *DNode
}

type DoubleFL struct {
    head *DNode
    tail *DNode
}

func createDoubleFL(beginNode string) DoubleFL {
    node := &DNode{value: beginNode}
    return DoubleFL{
        head: node,
        tail: node,
    }
}

func dblAddAtHead(doubleFlist *DoubleFL, data string) {
    if doubleFlist.head == nil {
        node := &DNode{value: data}
        doubleFlist.head = node
        doubleFlist.tail = node
        return
    }

    newHead := &DNode{value: data, next: doubleFlist.head}
    doubleFlist.head.prev = newHead
    doubleFlist.head = newHead
}

func dblAddAtTail(doubleFlist *DoubleFL, data string) {
    if doubleFlist.head == nil {
        node := &DNode{value: data}
        doubleFlist.head = node
        doubleFlist.tail = node
        return
    }

    newNode := &DNode{value: data, prev: doubleFlist.tail}
    doubleFlist.tail.next = newNode
    doubleFlist.tail = newNode
}

func dblAddBeforeIndex(doubleFlist *DoubleFL, data string, index int) {
    if doubleFlist.head == nil {
        fmt.Println("Список пуст.")
        return
    } else if index == 0 {
        dblAddAtHead(doubleFlist, data)
        return
    } else if index < 0 {
        fmt.Println("Неверный индекс!")
        return
    }

    i := 0
    curElem := doubleFlist.head

    for i != index && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil {
        fmt.Println("Элемента с таким индексом не существует!")
        return
    }

    newNode := &DNode{value: data, next: curElem, prev: curElem.prev}
    curElem.prev.next = newNode
    curElem.prev = newNode
}

func dblAddAfterIndex(doubleFlist *DoubleFL, data string, index int) {
    if doubleFlist.head == nil {
        fmt.Println("Список пуст.")
        return
    } else if index < 0 {
        fmt.Println("Неверный индекс!")
        return
    }

    curElem := doubleFlist.head
    i := 0

    for i != index && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil {
        fmt.Println("Элемента с таким индексом не существует!")
        return
    }

    if curElem == doubleFlist.tail {
        dblAddAtTail(doubleFlist, data)
        return
    }

    newNode := &DNode{value: data, next: curElem.next, prev: curElem}
    curElem.next.prev = newNode
    curElem.next = newNode
}

func dblDeleteAtHead(doubleFlist *DoubleFL) {
    if doubleFlist.head == nil {
        fmt.Println("Список пуст.")
        return
    }

    if doubleFlist.head == doubleFlist.tail {
        doubleFlist.head = nil
        doubleFlist.tail = nil
    } else {
        doubleFlist.head = doubleFlist.head.next
        doubleFlist.head.prev = nil
    }
}

func dblDeleteAtTail(doubleFlist *DoubleFL) {
    if doubleFlist.head == nil {
        fmt.Println("Список пуст.")
        return
    }

    if doubleFlist.head == doubleFlist.tail {
        doubleFlist.head = nil
        doubleFlist.tail = nil
    } else {
        doubleFlist.tail.prev.next = nil
        doubleFlist.tail = doubleFlist.tail.prev
    }
}

func dblDeleteBeforeIndex(doubleFlist *DoubleFL, index int) {
    if doubleFlist.head == nil || doubleFlist.head == doubleFlist.tail {
        fmt.Println("Удаление невозможно.")
        return
    } else if index == 1 {
        dblDeleteAtHead(doubleFlist)
        return
    } else if index <= 0 {
        fmt.Println("Неверный индекс.")
        return
    }

    curElem := doubleFlist.head
    i := 0

    for i != index && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil || curElem.prev == nil {
        fmt.Println("Неверный индекс.")
        return
    }

    deleteNode := curElem.prev

    if deleteNode == doubleFlist.head {
        dblDeleteAtHead(doubleFlist)
        return
    }

    deleteNode.prev.next = curElem
    curElem.prev = deleteNode.prev
}

func dblDeleteAfterIndex(doubleFlist *DoubleFL, index int) {
    if doubleFlist.head == nil || doubleFlist.head == doubleFlist.tail {
        fmt.Println("Удаление невозможно.")
        return
    } else if index < 0 {
        fmt.Println("Неверный индекс.")
        return
    }

    curElem := doubleFlist.head
    i := 0

    for i != index && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil || curElem == doubleFlist.tail {
        fmt.Println("Неверный индекс.")
        return
    }

    if curElem.next == doubleFlist.tail {
        dblDeleteAtTail(doubleFlist)
        return
    }

    deleteNode := curElem.next
    curElem.next = deleteNode.next
    deleteNode.next.prev = curElem
}

func dblDeleteValue(doubleFlist *DoubleFL, targetValue string) {
    if doubleFlist.head == nil {
        fmt.Println("Список пуст.")
        return
    }

    curElem := doubleFlist.head
    isFound := false

    for curElem != nil {
        if curElem.value == targetValue {
            isFound = true
            break
        }
        curElem = curElem.next
    }

    if !isFound {
        fmt.Println("Элемент не найден.")
        return
    }

    if curElem == doubleFlist.head {
        dblDeleteAtHead(doubleFlist)
        return
    }

    if curElem == doubleFlist.tail {
        dblDeleteAtTail(doubleFlist)
        return
    }

    deleteNode := curElem
    deleteNode.prev.next = deleteNode.next
    deleteNode.next.prev = deleteNode.prev
}

func dblSearchValue(doubleFlist *DoubleFL, targetValue string) {
    if doubleFlist.head == nil {
        fmt.Println("Список пуст.")
        return
    }

    curElem := doubleFlist.head
    isFound := false
    index := 0

    for curElem != nil {
        if curElem.value == targetValue {
            isFound = true
            break
        }
        curElem = curElem.next
        index++
    }

    if !isFound {
        fmt.Println("Элемент не найден.")
        return
    }

    fmt.Printf("Элемент %s найден. Индекс: %d\n", curElem.value, index)
}

func printDoubleFL(head *DNode) {
    fmt.Print("Двусвязный список: ")

    if head == nil {
        fmt.Println("Список пуст")
        return
    }

    for head != nil {
        fmt.Print(head.value + " ")
        head = head.next
    }

    fmt.Println()
}
